astroquery/mast/cloud.py

```diff
@@ -0,0 +1,191 @@
+        exception is raised. If the mission is supported but the product
+        cannot be found in the cloud, the returned path is None.
+
+        Parameters
+        ----------
+        data_product : `~astropy.table.Row`
+            Product to be converted into cloud data uri.
+        include_bucket : bool
+            When either to include the cloud bucket prefix in the result or not.
+        full_url : bool
+            Return a HTTP fetchable url instead of a uri.
+
+        Returns
+        -------
+        response : str or None
+            Cloud URI generated from the data product. If the product cannot be
+            found in the cloud, None is returned.
+        """
+
+        s3_client = self.boto3.client('s3')
+
+        paths = fpl.paths(data_product)
+        if paths is None:
+            raise Exception("Unsupported mission {}".format(data_product['obs_collection']))
+
+        for path in paths:
+            try:
+                s3_client.head_object(Bucket=self.pubdata_bucket, Key=path, RequestPayer='requester')
+                if include_bucket:
+                    path = "s3://{}/{}".format(self.pubdata_bucket, path)
+                elif full_url:
+                    path = "http://s3.amazonaws.com/{}/{}".format(self._pubdata_bucket, path)
+                return path
+            except self.botocore.exceptions.ClientError as e:
+                if e.response['Error']['Code'] != "404":
+                    raise
+
+        warnings.warn("Unable to locate file {}.".format(data_product['productFilename']), NoResultsWarning)
+        return None
+
+
+    def get_uri_list(self, data_products, include_bucket=True, full_url=False):
+        """
+        Takes an `~astropy.table.Table` of data products and returns the associated cloud data uris.
+
+        Parameters
+        ----------
+        data_products : `~astropy.table.Table`
+            Table containing products to be converted into cloud data uris.
+        include_bucket : bool
+            When either to include the cloud bucket prefix in the result or not.
+        full_url : bool
+            Return a HTTP fetchable url instead of a uri.
+
+        Returns
+        -------
+        response : list
+            List of URIs generated from the data products, list way contain entries that are None
+            if data_products includes products not found in the cloud.
+        """
+
+        return [self.get_uri(product, include_bucket, full_url) for product in data_products]
+
+    def download_file(self, data_product, local_path, cache=True):
+        """
+        Takes a data product in the form of an  `~astropy.table.Row` and downloads it from the cloud into
+        the given directory.
+
+        Parameters
+        ----------
+        data_product :  `~astropy.table.Row`
+            Product to download.
+        local_path : str
+            Directory in which files will be downloaded.
```

Comment:
local_path seems to be used as an actual file name, not a directory.  With caching it'll be important for the caller to know they're responsible for the uniqueness of the file name.